### ISSUE TITLE
Support For ItemFlow (SysML) and InformationFlow (UML)

### ISSUE THREAD

**Issue by danyeaw** [gaphor/gaphor#840]
**Is your feature request related to a problem? Please describe.**
In order to finish support for STPA, we need support for adding an ItemFlow to a Connector. This will also improve support for SysML and UML.

For example in SysML:
![image](https://user-images.githubusercontent.com/10014976/120083741-dd94a380-c098-11eb-9770-3b3cc3d51dc2.png)

The ItemFlow is represented by the black arrowhead on the connector. Torque is the itemProperty that belongs to the flow, and it is typed by a ValueType, Block, or Signal.

These are used in STPA to represent the flows in the control flow diagrams:
![image](https://user-images.githubusercontent.com/10014976/120082940-bab3c080-c093-11eb-9e1d-ce6efe92287c.png)

ItemFlows are also allowed on associations, for example creating a Context Diagram using a Block Definition Diagram.

UML also supports a similar diagram element called an InformationFlow with InformationItems representing the conveyed information that is flowing:
![image](https://user-images.githubusercontent.com/10014976/120083268-817c5000-c095-11eb-9046-d38b729777ff.png)

In addition to ValueType, Block, and Signal being allowed to represent the type of flow above, UML also supports other Classifiers like DataTypes and Signals.

**Describe the solution you'd like**
Add support for Signal elements with SysML and UML as well as InformationItems.

Drag a Block, Signal, DataType, ValueType, or InformationItem on top of a Connector and it creates an ItemFlow. Ideally dragging additional elements should be supported.

On the propertypage you can remove ItemFlows / InformationItems and change the flow direction.

**Describe alternatives you've considered**
Avoid usage of flows.

**Additional context**

**Comment by danyeaw** [comment]
Breaking down this feature into steps:

- [ ] Create a property page to handle adding and removing ItemFlow / itemProperty
- [ ] Add arrow style to Connectors when an ItemFlow exists on it
- [ ] Display any itemProperty that is part of an ItemFlow next to the Connector, including any stereotypes
- [ ] Maybe? expose the ItemFlow in the TreeView

@amolenaar do you have any thoughts on if this seems like a good order of implementation?

**Comment by amolenaar** [comment]
Do you have a reference to the specs? How does ItemFlow fit in the model?

**Comment by danyeaw** [comment]
Hey @amolenaar! ItemFlow is described in section 9.3.2.12 of the [SysML Spec](https://www.omg.org/spec/SysML/1.6/PDF).

**Comment by amolenaar** [comment]
I expected `ItemFlow` to be part of the meta model, but it looks like it's not defined there. I'm wondering how we should build it in Gaphor. Is it just a property on an `AssociationItem`/`ConnectorItem`? ItemFlow has an `itemProperty` attribute. I suppose it should be defined on the `Association`/`Connector` classes instead.

**Comment by amolenaar** [comment]
In this example:

![image](https://user-images.githubusercontent.com/10014976/120083268-817c5000-c095-11eb-9046-d38b729777ff.png)

I suppose the Company-Customer association would be implemented like this:

![image](https://user-images.githubusercontent.com/96249/126844543-34517582-cd4c-4bb5-baa4-e7a05a8c8a79.png)

**Comment by danyeaw** [comment]
@amolenaar Nice! I really like this visual representation. UML certainly made some of these concepts pretty complex!

**Comment by amolenaar** [comment]
I'll build it like this then, starting with Connector model element. We should find out what is the impact on the InformationFlow element when the Connector model connection is broken. Should the Connector generic function also manage the InformationFlow? Can we get away with making some associations composite?

**Comment by amolenaar** [comment]
@danyeaw Is this bit of editor enough for information/item flow?

![image](https://user-images.githubusercontent.com/96249/139919039-8b81fa52-d7c2-4c3b-8f30-51d17e9e3359.png)

**Comment by danyeaw** [comment]
@amolenaar How do we select the ItemFlow / itemProperty that exists in the model using the Property Page? If you type in the box does it add one?

**Comment by amolenaar** [comment]
I'm currently implementing it in terms of InformationFlow (the UML terminology).

It's a weird thing. InformationFlow in itself can be represented as a dependency:
![image](https://user-images.githubusercontent.com/96249/140419656-14c875fa-934a-4970-aeef-402cb7d80e75.png)

But the name shown is not the name of the InformationFlow, but of the InformationItem (or other classifier).

As for selecting the Informationitem, it's a text field now, but we may want to change it to a dropdown, so we can refer to existing classifiers.

**Comment by amolenaar** [comment]
With InformationFlow in, do we have enough support for now? Or should we add ItemFlow as well?

Maybe shortcut it by adding `itemProperty` to InformationFlow instead? If we add `itemProperty`, where should the properyt come from? The source or the target end of the ItemFlow/InformationFlow.

**Comment by danyeaw** [comment]
I think we only have InformationFlow enabled for SysML IBD connectors right? I think ItemFlow is just a more generally named version of InformationFlow, since for a system, you would want to flow other things besides information (like the torque example above).

Should we update InformationFlow to go on Associations between Classes, and replace InformationFlow with ItemFlow for SysML Connectors?

SysML says about itemProperty: An optional property that relates the flowing item to the instances of the connectors enclosing block. This 
property is applicable only for item flows realized by connectors. The itemProperty attribute has no values if the item flow is realized by an Association.

So if we have a Connector, we can create an instance of the ItemFlow type using an itemProperty. So if the ItemFlow type is Liquid, we can create an itemProperty called water that is an instance of Liquid.

It is a good question how we would add this property in Gaphor. Should it be another option in the Property page to add an instance of the type from the dropdown? This starts to get pretty complicated.

**Comment by amolenaar** [comment]
The Connector is an (ordinary) UML Connector.

I agree that it gets kinda complicated. Maybe leave it as is for now and focus on other elements instead?

**Comment by danyeaw** [comment]
I think the itemProperty is lower priority.

Isn't the SysML Connector the same as a UML Connector? Is there a challenge with using the ItemFlow? It seems like that would be higher priority to update.

**Comment by amolenaar** [comment]
We can use ItemFlow wherever we use InformationFlow atm. 

Maybe we can check of any of the connected elements are from the SysML package. If so, we use ItemFlow, else we use standard UML InformationFlow?

**Comment by amolenaar** [comment]
@danyeaw What would be the difference from a user point of view when using an `ItemFlow` instead of `InformationFlow`.

The SysML Connector is the same as the UML Connector.

**Comment by danyeaw** [comment]
I agree that to the user that there is no difference. However, if we only support connectors for SysML, shouldn't we use the correct element for the flow?

**Comment by amolenaar** [comment]
@danyeaw That would mean change all `InformationFlow` thingies to `ItemFlow` for now? I think that would make sense if we have a concrete use case for ItemFlow. If InformationFlow does suffice, I think it's better to leave it as is. Anyhow, ItemFlow is only applicable for Connectors. And InformationFlow is already way more advanced than what most people need, I assume.

**Comment by danyeaw** [comment]
@amolenaar You can add ItemFlows on Associations, just like an InformationFlow. I agree that what we have is great, but the use case is above with the STPA IBD for example, users are going to expect that the correctly named element if they are using SysML. The same thing could be said for Block vs. Class, sure a Block is a type of Class, but you wouldn't call it a Class when doing SysML modeling.

Do we want to add the ability to add them to associations as well?

**Comment by amolenaar** [comment]
@danyeaw I misinterpreted: indeed ItemFlow can be defined to an association, only the `itemProperty` remains empty. What does `itemProperty` refer to? A property from the source end? Or a separate property, owned by ItemFlow? I suppose the second.

**Comment by amolenaar** [comment]
@danyeaw I made a start with ItemFlow. All I'm struggling with is when to present InformationFlow and when to present ItemFlow.

**Comment by danyeaw** [comment]
@amolenaar Could it be based on the profile selected?

**Comment by amolenaar** [comment]
@danyeaw I do not have that information available. If this works, it should only select on profile if no InformationFlow/ItemFlow is set yet. If it's set, it should show the corresponding editor.

**Comment by amolenaar** [comment]
I think I found a solution.

Reasoning: It's confusing to have both the InformationFlow and the ItemFlow box showing in the property editor. However, since 2.7 is out, I can not just drop InformationFlow. And we may want it in a later version if we support connectors and stuff in UML.

What I made now is this: 
* the InformationFlow property page will only show when a connector has an InformationFlow set.
* The ItemFlow will always show, but the switch is insensitive when InformationFlow is set.

So to switch from InformationFlow to ItemFlow one has to:
* select the connector
* disable InformationFlow (at this point it can still be enabled)
* deselect the connector
* select it again
* enable ItemFlow (InformationFlow is no longer visible)

I do not want to make it more smart than that, to avoid weird conditions.
